packages/react-aria-components/src/Button.tsx

```diff
@@ -77,31 +109,98 @@ export const ButtonContext = createContext<ContextValue<ButtonContextValue, HTML
 
 function Button(props: ButtonProps, ref: ForwardedRef<HTMLButtonElement>) {
   [props, ref] = useContextProps(props, ref, ButtonContext);
+  props = disablePendingProps(props);
   let ctx = props as ButtonContextValue;
+  let {isPending, renderPendingState} = ctx;
   let {buttonProps, isPressed} = useButton(props, ref);
   let {focusProps, isFocused, isFocusVisible} = useFocusRing(props);
   let {hoverProps, isHovered} = useHover(props);
+  let renderValues = {isHovered, isPressed, isFocused, isFocusVisible, isDisabled: props.isDisabled || false, isPending};
   let renderProps = useRenderProps({
     ...props,
-    values: {isHovered, isPressed, isFocused, isFocusVisible, isDisabled: props.isDisabled || false},
+    values: renderValues,
     defaultClassName: 'react-aria-Button'
   });
 
+  let formatter = useLocalizedStringFormatter(intlMessages, 'react-aria-components');
+  // an aria label will block children and their labels from being read, this is undesirable for pending state
+  let hasAriaLabel = !!buttonProps['aria-label'] || !!buttonProps['aria-labelledby'];
+  let spinnerId = useId();
+  let buttonId = useId(buttonProps.id);
+  let containerId = useId();
+
+  const isPendingAriaLiveLabel = `${hasAriaLabel ? buttonProps['aria-label'] : ''} ${formatter.format('pending')}`.trim();
+  const isPendingAriaLiveLabelledby = hasAriaLabel ? (buttonProps['aria-labelledby']?.replace(buttonId, spinnerId) ?? spinnerId) : `${containerId} ${spinnerId}`.trim();
+
+  let ariaLive: 'off' | 'polite' | 'assertive' = 'polite';
+  if (isAppleDevice() && (!hasAriaLabel || isFirefox())) {
+    ariaLive = 'off';
+  }
+
   return (
     <button
       {...filterDOMProps(props, {propNames: additionalButtonHTMLAttributes})}
       {...mergeProps(buttonProps, focusProps, hoverProps)}
       {...renderProps}
+      id={buttonId}
       ref={ref}
+      aria-label={isPending ? isPendingAriaLiveLabel : buttonProps['aria-label']}
+      aria-labelledby={isPending ? isPendingAriaLiveLabelledby : buttonProps['aria-labelledby']}
       slot={props.slot || undefined}
+      aria-disabled={isPending ? 'true' : undefined}
       data-disabled={props.isDisabled || undefined}
       data-pressed={ctx.isPressed || isPressed || undefined}
       data-hovered={isHovered || undefined}
       data-focused={isFocused || undefined}
-      data-focus-visible={isFocusVisible || undefined} />
+      data-pending={isPending || undefined}
+      data-focus-visible={isFocusVisible || undefined}>
+      <div
+        style={{display: 'contents'}}
+        id={containerId}
+        aria-atomic>
+        {renderProps.children}
+      </div>
```

Comment:
As a user of RAC I'm a bit concerned about the overhead of 1 additional HTML element within each button. No matter if `isPending` is ever used or not. Buttons are a very frequently used component and it should be as slim as possible IMO.

Reply:
What are you worried about with the addition of one more HTML element? Is there a concrete issue? Buttons are already not what I would consider *slim*, see this three part blog series on building a button https://react-spectrum.adobe.com/blog/building-a-button-part-1.html

I'd love not to need these extra elements, but I haven't found a good way to handle the announcements without it so far.

Theoretically, buttons should just re-announce when you have focus on them and the contents change. However, this isn't true in practice. https://a11ysupport.io/tests/tech__html__button-name-change and even the ones that say they have support have different kinds of support.

When the button enters the pending state, we want that to be the first thing read. For an element which can appear and disappear, I think it makes more sense to keep it as the second sibling though so it doesn't push the original label out of the way if someone wants to keep the original label visible.
Someone could still visually change that with flex order, however, I didn't want to impose that on all buttons unexpectedly.

In order to get everything announced in the right order without double announcing, we use aria-atomic so the screen reader doesn't read off anything past the content except what we wish by joining together aria-labelledby.

We'll keep thinking on it and thanks for the feedback.

Reply:
I just remember that you should keep the amount of DOM nodes as low as possible. I think that's especially important in such a base building block like a component library/design system so you have room on application-level.